File: 23_Checkers/javascript/checkers.js
// CHECKERS
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// x,y = origin square
// a,b = movement direction
function try_computer()
{
    u = x + a;
    v = y + b;
    if (u < 0 || u > 7 || v < 0 || v > 7)
        return;
    if (s[u][v] == 0) {
        eval_move();
        return;
    }
    if (s[u][v] < 0)	// Cannot jump over own pieces
        return;
    u += a;
    u += b;
    if (u < 0 || u > 7 || v < 0 || v > 7)
        return;
    if (s[u][v] == 0)
        eval_move();
}

// x,y = origin square
// u,v = target square
function eval_move()
{
    if (v == 0 && s[x][y] == -1)
        q += 2;
    if (Math.abs(y - v) == 2)
        q += 5;
    if (y == 7)
        q -= 2;
    if (u == 0 || u == 7)
        q++;
    for (c = -1; c <= 1; c += 2) {
        if (u + c < 0 || u + c > 7 || v + g < 0)
            continue;
        if (s[u + c][v + g] < 0) {	// Computer piece
            q++;
            continue;
        }
        if (u - c < 0 || u - c > 7 || v - g > 7)
            continue;
        if (s[u + c][v + g] > 0 && (s[u - c][v - g] == 0 || (u - c == x && v - g == y)))
            q -= 2;
    }
    if (q > r[0]) {	// Best movement so far?
        r[0] = q;	// Take note of score
        r[1] = x;	// Origin square
        r[2] = y;
        r[3] = u;	// Target square
        r[4] = v;
    }
    q = 0;
}

function more_captures() {
    u = x + a;
    v = y + b;
    if (u < 0 || u > 7 || v < 0 || v > 7)
        return;
    if (s[u][v] == 0 && s[x + a / 2][y + b / 2] > 0)
        eval_move();
}

var r = [-99, 0, 0, 0, 0];
var s = [];

for (x = 0; x <= 7; x++)
    s[x] = [];
    
var g = -1;
var data = [1, 0, 1, 0, 0, 0, -1, 0, 0, 1, 0, 0, 0, -1, 0, -1, 15];
var p = 0;
var q = 0;

// Main program
async function main()
{
    print(tab(32) + "CHECKERS\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("THIS IS THE GAME OF CHECKERS.  THE COMPUTER IS X,\n");
    print("AND YOU ARE O.  THE COMPUTER WILL MOVE FIRST.\n");
    print("SQUARES ARE REFERRED TO BY A COORDINATE SYSTEM.\n");
    print("(0,0) IS THE LOWER LEFT CORNER\n");
    print("(0,7) IS THE UPPER LEFT CORNER\n");
    print("(7,0) IS THE LOWER RIGHT CORNER\n");
    print("(7,7) IS THE UPPER RIGHT CORNER\n");
    print("THE COMPUTER WILL TYPE '+TO' WHEN YOU HAVE ANOTHER\n");
    print("JUMP.  TYPE TWO NEGATIVE NUMBERS IF YOU CANNOT JUMP.\n");
    print("\n");
    print("\n");
    print("\n");
    for (x = 0; x <= 7; x++) {
        for (y = 0; y <= 7; y++) {
            if (data[p] == 15)
                p = 0;
            s[x][y] = data[p];
            p++;
        }
    }
    while (1) {

        // Search the board for the best movement
        for (x = 0; x <= 7; x++) {
            for (y = 0; y <= 7; y++) {
                if (s[x][y] > -1)
                    continue;
                if (s[x][y] == -1) {	// Piece
                    for (a = -1; a <= 1; a += 2) {
                        b = g;	// Only advances
                        try_computer();
                    }
                } else if (s[x][y] == -2) {	// King
                    for (a = -1; a <= 1; a += 2) {
                        for (b = -1; b <= 1; b += 2) {
                            try_computer();
                        }
                    }
                }
            }
        }
        if (r[0] == -99) {
            print("\n");
            print("YOU WIN.\n");
            break;
        }
        print("FROM " + r[1] + "," + r[2] + " TO " + r[3] + "," + r[4]);
        r[0] = -99;
        while (1) {
            if (r[4] == 0) {	// Computer reaches the bottom
                s[r[3]][r[4]] = -2;	// King
                break;
            }
            s[r[3]][r[4]] = s[r[1]][r[2]];	// Move
            s[r[1]][r[2]] = 0;
            if (Math.abs(r[1] - r[3]) == 2) {
                s[(r[1] + r[3]) / 2][(r[2] + r[4]) / 2] = 0;	// Capture
                x = r[3];
                y = r[4];
                if (s[x][y] == -1) {
                    b = -2;
                    for (a = -2; a <= 2; a += 4) {
                        more_captures();
                    }
                } else if (s[x][y] == -2) {
                    for (a = -2; a <= 2; a += 4) {
                        for (b = -2; b <= 2; b += 4) {
                            more_captures();
                        }
                    }
                }
                if (r[0] != -99) {
                    print(" TO " + r[3] + "," + r[4]);
                    r[0] = -99;
                    continue;
                }
            }
            break;
        }
        print("\n");
        print("\n");
        print("\n");
        for (y = 7; y >= 0; y--) {
            str = "";
            for (x = 0; x <= 7; x++) {
                if (s[x][y] == 0)
                    str += ".";
                if (s[x][y] == 1)
                    str += "O";
                if (s[x][y] == -1)
                    str += "X";
                if (s[x][y] == -2)
                    str += "X*";
                if (s[x][y] == 2)
                    str += "O*";
                while (str.length % 5)
                    str += " ";
            }
            print(str + "\n");
            print("\n");
        }
        print("\n");
        z = 0;
        t = 0;
        for (l = 0; l <= 7; l++) {
            for (m = 0; m <= 7; m++) {
                if (s[l][m] == 1 || s[l][m] == 2)
                    z = 1;
                if (s[l][m] == -1 || s[l][m] == -2)
                    t = 1;
            }
        }
        if (z != 1) {
            print("\n");
            print("I WIN.\n");
            break;
        }
        if (t != 1) {
            print("\n");
            print("YOU WIN.\n");
            break;
        }
        do {
            print("FROM");
            e = await input();
            h = parseInt(e.substr(e.indexOf(",") + 1));
            e = parseInt(e);
            x = e;
            y = h;
        } while (s[x][y] <= 0) ;
        do {
            print("TO");
            a = await input();
            b = parseInt(a.substr(a.indexOf(",") + 1));
            a = parseInt(a);
            x = a;
            y = b;
            if (s[x][y] == 0 && Math.abs(a - e) <= 2 && Math.abs(a - e) == Math.abs(b - h))
                break;
            print("WHAT?\n");
        } while (1) ;
        i = 46;
        do {
            s[a][b] = s[e][h]
            s[e][h] = 0;
            if (Math.abs(e - a) != 2)
                break;
            s[(e + a) / 2][(h + b) / 2] = 0;
            while (1) {
                print("+TO");
                a1 = await input();
                b1 = parseInt(a1.substr(a1.indexOf(",") + 1));
                a1 = parseInt(a1);
                if (a1 < 0)
                    break;
                if (s[a1][b1] == 0 && Math.abs(a1 - a) == 2 && Math.abs(b1 - b) == 2)
                    break;
            }
            if (a1 < 0)
                break;
            e = a;
            h = b;
            a = a1;
            b = b1;
            i += 15;
        } while (1);
        if (b == 7)	// Player reaches top
            s[a][b] = 2;	// Convert to king
    }
}

main();

File: 23_Checkers/python/checkers.py
"""
CHECKERS

How about a nice game of checkers?

Ported by Dave LeCompte
"""

from typing import Iterator, NamedTuple, Optional, Tuple

PAGE_WIDTH = 64

HUMAN_PLAYER = 1
COMPUTER_PLAYER = -1
HUMAN_PIECE = 1
HUMAN_KING = 2
COMPUTER_PIECE = -1
COMPUTER_KING = -2
EMPTY_SPACE = 0

TOP_ROW = 7
BOTTOM_ROW = 0


class MoveRecord(NamedTuple):
    quality: int
    start_x: int
    start_y: int
    dest_x: int
    dest_y: int


def print_centered(msg: str) -> None:
    spaces = " " * ((PAGE_WIDTH - len(msg)) // 2)
    print(spaces + msg)


def print_header(title: str) -> None:
    print_centered(title)
    print_centered("CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print()
    print()
    print()


def get_coordinates(prompt: str) -> Tuple[int, int]:
    err_msg = "ENTER COORDINATES in X,Y FORMAT"
    while True:
        print(prompt)
        response = input()
        if "," not in response:
            print(err_msg)
            continue

        try:
            x, y = (int(c) for c in response.split(","))
        except ValueError:
            print(err_msg)
            continue

        return x, y


def is_legal_board_coordinate(x: int, y: int) -> bool:
    return (0 <= x <= 7) and (0 <= y <= 7)


class Board:
    def __init__(self) -> None:
        self.spaces = [[0 for y in range(8)] for x in range(8)]
        for x in range(8):
            if (x % 2) == 0:
                self.spaces[x][6] = COMPUTER_PIECE
                self.spaces[x][2] = HUMAN_PIECE
                self.spaces[x][0] = HUMAN_PIECE
            else:
                self.spaces[x][7] = COMPUTER_PIECE
                self.spaces[x][5] = COMPUTER_PIECE
                self.spaces[x][1] = HUMAN_PIECE

    def __str__(self) -> str:
        pieces = {
            EMPTY_SPACE: ".",
            HUMAN_PIECE: "O",
            HUMAN_KING: "O*",
            COMPUTER_PIECE: "X",
            COMPUTER_KING: "X*",
        }

        s = "\n\n\n"
        for y in range(7, -1, -1):
            for x in range(0, 8):
                piece_str = pieces[self.spaces[x][y]]
                piece_str += " " * (5 - len(piece_str))
                s += piece_str
            s += "\n"
        s += "\n\n"

        return s

    def get_spaces(self) -> Iterator[Tuple[int, int]]:
        for x in range(0, 8):
            for y in range(0, 8):
                yield x, y

    def get_spaces_with_computer_pieces(self) -> Iterator[Tuple[int, int]]:
        for x, y in self.get_spaces():
            contents = self.spaces[x][y]
            if contents < 0:
                yield x, y

    def get_spaces_with_human_pieces(self) -> Iterator[Tuple[int, int]]:
        for x, y in self.get_spaces():
            contents = self.spaces[x][y]
            if contents > 0:
                yield x, y

    def get_legal_deltas_for_space(self, x: int, y: int) -> Iterator[Tuple[int, int]]:
        contents = self.spaces[x][y]
        if contents == COMPUTER_PIECE:
            for delta_x in (-1, 1):
                yield (delta_x, -1)
        else:
            for delta_x in (-1, 1):
                for delta_y in (-1, 1):
                    yield (delta_x, delta_y)

    def get_legal_moves(self, x: int, y: int) -> Iterator[MoveRecord]:
        for delta_x, delta_y in self.get_legal_deltas_for_space(x, y):
            new_move_record = self.check_move(x, y, delta_x, delta_y)

            if new_move_record is not None:
                yield new_move_record

    def pick_computer_move(self) -> Optional[MoveRecord]:
        move_record = None

        for start_x, start_y in self.get_spaces_with_computer_pieces():
            for delta_x, delta_y in self.get_legal_deltas_for_space(start_x, start_y):
                new_move_record = self.check_move(start_x, start_y, delta_x, delta_y)

                if new_move_record is None:
                    continue

                if (move_record is None) or (
                    new_move_record.quality > move_record.quality
                ):
                    move_record = new_move_record

        return move_record

    def check_move(
        self, start_x: int, start_y: int, delta_x: int, delta_y: int
    ) -> Optional[MoveRecord]:
        new_x = start_x + delta_x
        new_y = start_y + delta_y
        if not is_legal_board_coordinate(new_x, new_y):
            return None

        contents = self.spaces[new_x][new_y]
        if contents == EMPTY_SPACE:
            return self.evaluate_move(start_x, start_y, new_x, new_y)
        if contents < 0:
            return None

        # check jump landing space, which is an additional dx, dy from new_x, newy
        landing_x = new_x + delta_x
        landing_y = new_y + delta_y

        if not is_legal_board_coordinate(landing_x, landing_y):
            return None
        if self.spaces[landing_x][landing_y] == EMPTY_SPACE:
            return self.evaluate_move(start_x, start_y, landing_x, landing_y)
        return None

    def evaluate_move(
        self, start_x: int, start_y: int, dest_x: int, dest_y: int
    ) -> MoveRecord:
        quality = 0
        if dest_y == 0 and self.spaces[start_x][start_y] == COMPUTER_PIECE:
            # promoting is good
            quality += 2
        if abs(dest_y - start_y) == 2:
            # jumps are good
            quality += 5
        if start_y == 7:
            # prefer to defend back row
            quality -= 2
        if dest_x in (0, 7):
            # moving to edge column
            quality += 1
        for delta_x in (-1, 1):
            if not is_legal_board_coordinate(dest_x + delta_x, dest_y - 1):
                continue

            if self.spaces[dest_x + delta_x][dest_y - 1] < 0:
                # moving into "shadow" of another computer piece
                quality += 1

            if not is_legal_board_coordinate(dest_x - delta_x, dest_y + 1):
                continue

            if (
                (self.spaces[dest_x + delta_x][dest_y - 1] > 0)
                and (self.spaces[dest_x - delta_x][dest_y + 1] == EMPTY_SPACE)
                or ((dest_x - delta_x == start_x) and (dest_y + 1 == start_y))
            ):
                # we are moving up to a human checker that could jump us
                quality -= 2
        return MoveRecord(quality, start_x, start_y, dest_x, dest_y)

    def remove_r_pieces(self, move_record: MoveRecord) -> None:
        self.remove_pieces(
            move_record.start_x,
            move_record.start_y,
            move_record.dest_x,
            move_record.dest_y,
        )

    def remove_pieces(
        self, start_x: int, start_y: int, dest_x: int, dest_y: int
    ) -> None:
        self.spaces[dest_x][dest_y] = self.spaces[start_x][start_y]
        self.spaces[start_x][start_y] = EMPTY_SPACE

        if abs(dest_x - start_x) == 2:
            mid_x = (start_x + dest_x) // 2
            mid_y = (start_y + dest_y) // 2
            self.spaces[mid_x][mid_y] = EMPTY_SPACE

    def play_computer_move(self, move_record: MoveRecord) -> None:
        print(
            f"FROM {move_record.start_x} {move_record.start_y} TO {move_record.dest_x} {move_record.dest_y}"
        )

        while True:
            if move_record.dest_y == BOTTOM_ROW:
                # KING ME
                self.remove_r_pieces(move_record)
                self.spaces[move_record.dest_x][move_record.dest_y] = COMPUTER_KING
                return
            else:
                self.spaces[move_record.dest_x][move_record.dest_y] = self.spaces[
                    move_record.start_x
                ][move_record.start_y]
                self.remove_r_pieces(move_record)

                if abs(move_record.dest_x - move_record.start_x) != 2:
                    return

                landing_x = move_record.dest_x
                landing_y = move_record.dest_y

                best_move = None
                if self.spaces[landing_x][landing_y] == COMPUTER_PIECE:
                    for delta_x in (-2, 2):
                        test_record = self.try_extend(landing_x, landing_y, delta_x, -2)
                        if (move_record is not None) and (
                            (best_move is None)
                            or (move_record.quality > best_move.quality)
                        ):
                            best_move = test_record
                else:
                    assert self.spaces[landing_x][landing_y] == COMPUTER_KING
                    for delta_x in (-2, 2):
                        for delta_y in (-2, 2):
                            test_record = self.try_extend(
                                landing_x, landing_y, delta_x, delta_y
                            )
                            if (move_record is not None) and (
                                (best_move is None)
                                or (move_record.quality > best_move.quality)
                            ):
                                best_move = test_record

                if best_move is None:
                    return
                else:
                    print(f"TO {best_move.dest_x} {best_move.dest_y}")
                    move_record = best_move

    def try_extend(
        self, start_x: int, start_y: int, delta_x: int, delta_y: int
    ) -> Optional[MoveRecord]:
        new_x = start_x + delta_x
        new_y = start_y + delta_y

        if not is_legal_board_coordinate(new_x, new_y):
            return None

        jumped_x = start_x + delta_x // 2
        jumped_y = start_y + delta_y // 2

        if (self.spaces[new_x][new_y] == EMPTY_SPACE) and (
            self.spaces[jumped_x][jumped_y] > 0
        ):
            return self.evaluate_move(start_x, start_y, new_x, new_y)
        return None

    def get_human_move(self) -> Tuple[int, int, int, int]:
        is_king = False

        while True:
            start_x, start_y = get_coordinates("FROM?")

            legal_moves = list(self.get_legal_moves(start_x, start_y))
            if not legal_moves:
                print(f"({start_x}, {start_y}) has no legal moves. Choose again.")
                continue
            if self.spaces[start_x][start_y] > 0:
                break

        is_king = self.spaces[start_x][start_y] == HUMAN_KING

        while True:
            dest_x, dest_y = get_coordinates("TO?")

            if (not is_king) and (dest_y < start_y):
                # CHEATER! Trying to move non-king backwards
                continue
            is_free = self.spaces[dest_x][dest_y] == 0
            within_reach = abs(dest_x - start_x) <= 2
            is_diagonal_move = abs(dest_x - start_x) == abs(dest_y - start_y)
            if is_free and within_reach and is_diagonal_move:
                break
        return start_x, start_y, dest_x, dest_y

    def get_human_extension(
        self, start_x: int, start_y: int
    ) -> Tuple[bool, Optional[Tuple[int, int, int, int]]]:
        is_king = self.spaces[start_x][start_y] == HUMAN_KING

        while True:
            dest_x, dest_y = get_coordinates("+TO?")

            if dest_x < 0:
                return False, None
            if (not is_king) and (dest_y < start_y):
                # CHEATER! Trying to move non-king backwards
                continue
            if (
                (self.spaces[dest_x][dest_y] == EMPTY_SPACE)
                and (abs(dest_x - start_x) == 2)
                and (abs(dest_y - start_y) == 2)
            ):
                return True, (start_x, start_y, dest_x, dest_y)

    def play_human_move(
        self, start_x: int, start_y: int, dest_x: int, dest_y: int
    ) -> None:
        self.remove_pieces(start_x, start_y, dest_x, dest_y)

        if dest_y == TOP_ROW:
            # KING ME
            self.spaces[dest_x][dest_y] = HUMAN_KING

    def check_pieces(self) -> bool:
        if len(list(self.get_spaces_with_computer_pieces())) == 0:
            print_human_won()
            return False
        if len(list(self.get_spaces_with_computer_pieces())) == 0:
            print_computer_won()
            return False
        return True


def print_instructions() -> None:
    print("THIS IS THE GAME OF CHECKERS.  THE COMPUTER IS X,")
    print("AND YOU ARE O.  THE COMPUTER WILL MOVE FIRST.")
    print("SQUARES ARE REFERRED TO BY A COORDINATE SYSTEM.")
    print("(0,0) IS THE LOWER LEFT CORNER")
    print("(0,7) IS THE UPPER LEFT CORNER")
    print("(7,0) IS THE LOWER RIGHT CORNER")
    print("(7,7) IS THE UPPER RIGHT CORNER")
    print("THE COMPUTER WILL TYPE '+TO' WHEN YOU HAVE ANOTHER")
    print("JUMP.  TYPE TWO NEGATIVE NUMBERS IF YOU CANNOT JUMP.")
    print()
    print()
    print()


def print_human_won() -> None:
    print()
    print("YOU WIN.")


def print_computer_won() -> None:
    print()
    print("I WIN.")


def play_game() -> None:
    board = Board()

    while True:
        move_record = board.pick_computer_move()
        if move_record is None:
            print_human_won()
            return
        board.play_computer_move(move_record)

        print(board)

        if not board.check_pieces():
            return

        start_x, start_y, dest_x, dest_y = board.get_human_move()
        board.play_human_move(start_x, start_y, dest_x, dest_y)
        if abs(dest_x - start_x) == 2:
            while True:
                extend, move = board.get_human_extension(dest_x, dest_y)
                assert move is not None
                if not extend:
                    break
                start_x, start_y, dest_x, dest_y = move
                board.play_human_move(start_x, start_y, dest_x, dest_y)


def main() -> None:
    print_header("CHECKERS")
    print_instructions()

    play_game()


if __name__ == "__main__":
    main()
